contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -2867,6 +2879,377 @@ pltsql_detect_numeric_overflow(int weight, int dscale, int first_block, int nume
+				field[0] = res->data;
+				ftype[0] = DTK_DATE;
+				nf = nf - 2;
+
+				/*
+				 * Since the first 3 fields converted to 1, skip the
+				 * attached fields i.e., field[1]/[2], ftype[1]/[2]
+				 */
+				for(int i = 1; i < nf; i++)
+				{
+					field[i] = field[i+2];
+					ftype[i] = ftype[i+2];
+				}
+			}
+		}
+
+		for (int i = 0; i < nf; i++)
+		{
+			int len;
+			char* different_date_formats[] = {"/", ".", "-"};
+			temp_field = pstrdup(field[i]);
+			len = strlen(temp_field);
+
+			switch (ftype[i])
+			{
+				case DTK_NUMBER:
+					/* If only number fields are present in
+					 * the input then following considerations should
+					 * be followed.
+					 * 1. len = 1,2 : Time and [ap]m should be specified
+					 * 2. len = 4 : Considered as year with format "YYYY-01-01"
+					 * 3. len = 6 : Considered as format `"YYMMDD"
+					 * 4. len = 8 : Considered as format "YYYYMMDD"
+					 * 5. len = 3,5,7,>8 :  Not an valid and should throw error
+					 */
+					switch (len)
+					{
+						case 1:
+						case 2:
+							/*
+							 * For example if the input is "1 am"
+							 * will be converted to "1:00:00 am".
+							 */
+							if (i == nf - 1 || ftype[i+1] != DTK_STRING)
+								TIME_IN_ERROR();
+							if (pg_strcasecmp(field[i+1], "am") == 0 || pg_strcasecmp(field[i+1], "pm") == 0)
+							{
+								res = makeStringInfo();
+								appendStringInfo(res, "%s%s", field[i], ":00:00");
+								field[i] = res->data;
+								ftype[i] = DTK_TIME;
+							}
+							else
+								TIME_IN_ERROR();
+							break;
+
+						case 4:
+							/*
+							 * For example if input is "2000" will be converted
+							 * to "2000-01-01".
+							 */
+							res = makeStringInfo();
+							appendStringInfo(res, "%s%s", field[i], "-01-01");
+							field[i] = res->data;
+							ftype[i] = DTK_DATE;
+							break;
+
+						case 6:
+						case 8:
+							/*
+							 * For example if input is "201213" will be converted
+							 * to "2000-12-13", and input "20001118" will be
+							 * converted to "2000-11-18"
+							 */
+							res = makeStringInfo();
+							for (int k = 0; k < len; k++)
+							{
+								appendStringInfo(res, "%c", temp_field[k]);
+								if ((len == 6 && ((k + 1) % 2 == 0 && k < 5)) ||
+										(len == 8 && (k == 3 || k == 5)))
+									appendStringInfo(res, "%c", '-');
+							}
+							field[i] = res->data;	
+							ftype[i] = DTK_DATE;
+							break;
+
+						default:
+							TIME_IN_ERROR();
+					}
+					break;
+
+				case DTK_DATE:
+					/*
+					 * If the input is of format '<DATE>T<TIME>, then the
+					 * <DATE> should be of format "YYYY-MM-DD"
+					 */
+					if (i == 0 && nf > 1 && field[1] && pg_strcasecmp(field[1], "t") == 0)
+					{
+						pattern = "^[1-9][0-9]{3}-[0-9]?[0-9]-[0-9]?[0-9]$";
+						if (regcomp(&time_regex, pattern, REG_EXTENDED) != 0)
+							ereport(ERROR,
+										(errcode(ERRCODE_INTERNAL_ERROR),
+										errmsg("time format regex failed")));
```

Comment:
we should just say time format internal error

Reply:
How about, we just send the same message as conversion failed?